eurovision/notebooks/story.ipynb

```diff
@@ -2090,7 +2039,269 @@
+    "Thus, any year where the UK performs well is a surprise to our model!\n",
+    "\n",
+    "Having looked at how good our model is at reproducing its own data, let's see what happens when we predict the 2023 results (now including all years' worth of data in our training)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "predictions_2023 = predict_baseline(train_baseline(), 2023)\n",
+    "\n",
+    "n = len(predictions_2023)\n",
+    "\n",
+    "fig, ax = plt.subplots(figsize=(7, 7))\n",
+    "ax.barh(width=predictions_2023['rescaled_rank'], y=np.arange(n), color='#4878D0')\n",
+    "ax.set(\n",
+    "    xlim=(0, predictions_2023['rescaled_rank'].max() * 1.1),\n",
+    "    ylim=(-1, n),\n",
+    "    yticks=np.arange(n),\n",
+    "    yticklabels=predictions_2023.index.values,\n",
+    "    xlabel='Average rescaled rank across all final appearances',\n",
+    "    title='Baseline model: predicting 2023 results based on historical performance'\n",
+    ")\n",
+    "ax.invert_yaxis()\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Our baseline model thus predicts Italy to come first, Sweden second, and Ukraine third, simply on the basis of them performing historically well at Eurovision.\n",
+    "\n",
+    "With the benefit of hindsight, we can calculate the Spearman correlation coefficient for 2023.\n",
+    "Since we cannot really assign a 'rank' to contestants that were eliminated in the semi-finals, this coefficient is calculated only for those countries which made the finals."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "actual_ranks_2023 = (future[~np.isnan(future['rank'])].drop_duplicates(subset=['to_code2']))[['to_code2', 'rank']].set_index('to_code2')\n",
+    "\n",
+    "finalists_2023 = actual_ranks_2023.index.unique()\n",
+    "predictions_2023_finals = predictions_2023.query('to_code2 in @finalists_2023')\n",
+    "\n",
+    "print(f'Spearman correlation for 2023: {get_spearman(predictions_2023_finals, actual_ranks_2023):.3f}')\n",
+    "\n",
+    "both_2023 = join_predictions_and_actual(predictions_2023_finals, actual_ranks_2023)\n",
+    "ax = both_2023.plot(kind='scatter', x='predicted_rank', y='actual_rank', figsize=(4, 4), color='#6495ed')\n",
+    "ax.set(xlabel='Predicted rank (by baseline model)', ylabel='Actual rank', title='Baseline model: 2023 predictions vs. actual results')\n",
+    "for i, txt in enumerate(both_2023.index):\n",
+    "    ax.annotate(txt, (both_2023.iloc[i]['predicted_rank'], both_2023.iloc[i]['actual_rank'] - 0.2),\n",
+    "                ha='center', va='bottom', fontsize=8)\n",
+    "ax.invert_xaxis(); ax.invert_yaxis()\n",
+    "plt.show()"
```

Comment:
It might be harder to have country names here (it may ruin the clarity of the graph), but might be worth trying just in case.

Reply:
[`pip install adjustText`](https://adjusttext.readthedocs.io/en/latest/) to the rescue ...

<img width="423" alt="Screenshot 2023-06-29 at 00 16 56" src="https://github.com/KatrionaGoldmann/Eurovision_TDS/assets/22414895/4fadaac8-4b0f-46be-bd55-6cd99ab2abf1">

Reply:
Oh, yes, nice solution. I do think that's much easier to understand for the reader.